nemo/collections/asr/models/ctc_models.py

```diff
@@ -524,6 +536,23 @@ def forward(
         encoded_len = encoder_output[1]
         log_probs = self.decoder(encoder_output=encoded)
         greedy_predictions = log_probs.argmax(dim=-1, keepdim=False)
+        # generating decoding results for intermediate layers if necessary
+        if self.intermediate_loss_weights:
+            # we assume that encoder has to have property called "captured_layer_outputs"
+            # which is a list with the same length as loss weights
+            if len(self.encoder.captured_layer_outputs) != len(self.intermediate_loss_weights):
```

Comment:
Dont assume. First and foremost you should not be attaching tensors with grads to a module - ever. You can instead use the Tensor Registry framework in NeMo. Most models already support it due to other parts needing it. So do the following - In the training_step, enable the tensor registry, let modules register their forward activation (or a subset of it) and then access the registry here. Empty the registry at end of the train step to prevent memory leak.

So do a hasattr check, then raise a proper error (Value/Runtime) which details 
1) Provided CTC intermediate loss weights list is not empty
2) But the model does not add anything to the registry.

Also, this will waste memory during inference if the user does not need interctc outputs unless explicitly requested. So put a self.training check above for this part.